Doc/library/__main__.rst

```diff
@@ -1,25 +1,182 @@
-standard input, a script, or from an interactive prompt.
+In Python, ``__main__`` is not a single mechanism in the language, but in fact
+is part of two quite different constructs:
+
+1. The ``__name__ == '__main__'`` statement
+2. The ``__main__.py`` file in Python packages
+
+Each of these mechanisms are related to Python modules; how users interact with
+them and how they interact with each other. See section :ref:`tut-modules`.
+
+
+.. _name_is_main:
+
+``__name__ == '__main__'``
+---------------------------
+
+``'__main__'`` is the name of the environment where top-level code is run.
+"Top-level code" means when a Python module is initialized from an interactive
+prompt, from standard input, from a file argument, from a :option:`-c`
+argument, or from a :option:`-m` argument, but **not** when it is initialized
+from an import statement.  In any of these situations, the module's
+``__name__`` is set equal to ``'__main__'``.
+
+The only other context in which Python code is run is when it is imported
+through an import statement. In that case, ``__name__`` is set equal to the
+module's name: usually the name of the file without the ``.py`` extension.
+
+As a result, a module can discover whether or not it is running in the
+top-level environment by checking its own ``__name__``, which allows a common
+idiom for conditionally executing code when the module is not initialized from
+an import statement::
+
+    if __name__ == '__main__':
+        # Execute when the module is not initialized from an import statement.
+        ...
+
+
+Idiomatic Usage
+^^^^^^^^^^^^^^^
+
+Putting as few statements as possible in the block below ``if __name___ ==
+'__main__'`` can improve code clarity. Most often, a function named *main*
+encapsulates the program's primary behavior::
+
+    # echo.py
+
+    import shlex
+    import sys
+
+
+    def echo(phrase: str):
+       """A dummy wrapper around print."""
+       # for demonstration purposes, you can imagine that there is some
+       # valuable and reusable logic inside this function
+       print(phrase)
+
+
+    def main():
+        """Echo the input arguments to standard output"""
+        echo(shlex.join(sys.argv))
+
+
+    if __name__ == '__main__':
+        sys.exit(main())  # next section explains the use of sys.exit
+
+This has the added benefit of the *echo* function itself being isolated and
+importable elsewhere. None of the code in ``echo.py`` will execute at
+import-time.
+
+
+Packaging Considerations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For detailed documentation on Python packaging, see the
+`Python Packaging User Guide. <https://packaging.python.org/>`_
+
+*main* functions are often used to create command line tools by specifying them
+as entry points for console scripts.  When this is done, pip inserts the
+function call into a template script, where the return value of *main* is
+passed into sys.exit. For example::
```

Comment:
```suggestion
passed into ``sys.exit``. For example::
```
Or `` :func:`sys.exit` ``, whichever you prefer.

Reply:
Definitely the latter, but thank you for the catch!